components/Ink/src/private/MDCInkLayer.m

```diff
@@ -446,57 +553,60 @@ - (void)spreadFromPoint:(CGPoint)point completion:(void (^)())completionBlock {
-  _foregroundRipple.bounded = self.bounded;
-  _foregroundRipple.animationDelegate = self;
-  _foregroundRipple.useCustomInkCenter = self.useCustomInkCenter;
-  _foregroundRipple.customInkCenter = self.customInkCenter;
-  [_foregroundRipple setupRipple];
-
-  [_backgroundRipple enter];
-  [_foregroundRipple enter];
-
-  _spreadCompletionBlock = completionBlock;
+  MDCInkLayerBackgroundRipple *backgroundRipple = [[MDCInkLayerBackgroundRipple alloc] init];
+  backgroundRipple.inkLayer = _compositeRipple;
+  backgroundRipple.targetFrame = self.bounds;
+  backgroundRipple.point = point;
+  backgroundRipple.color = self.inkColor;
+  backgroundRipple.radius = radius;
+  backgroundRipple.bounded = self.bounded;
+  backgroundRipple.animationDelegate = self;
+  [backgroundRipple setupRipple];
+
+  MDCInkLayerForegroundRipple *foregroundRipple = [[MDCInkLayerForegroundRipple alloc] init];
+  foregroundRipple.inkLayer = _compositeRipple;
+  foregroundRipple.targetFrame = self.bounds;
+  foregroundRipple.point = point;
+  foregroundRipple.color = self.inkColor;
+  foregroundRipple.radius = radius;
+  foregroundRipple.bounded = self.bounded;
+  foregroundRipple.animationDelegate = self;
+  foregroundRipple.useCustomInkCenter = self.useCustomInkCenter;
+  foregroundRipple.customInkCenter = self.customInkCenter;
+  [foregroundRipple setupRipple];
+
+  [backgroundRipple enter];
+  [self.backgroundRipples addObject:backgroundRipple];
+  [foregroundRipple enterWithCompletion:completionBlock];
+  [self.foregroundRipples addObject:foregroundRipple];
+  self.unexitedForegroundRippleIndex++;
+  self.unexitedBackgroundRippleIndex++;
+
 }
 
 - (void)evaporateWithCompletion:(void (^)())completionBlock {
-  _evaporateCompletionBlock = completionBlock;
-  [self reset:YES];
+  [self resetBottomInk:YES completion:completionBlock];
 }
 
 - (void)evaporateToPoint:(CGPoint)point completion:(void (^)())completionBlock {
-  _evaporateToPointCompletionBlock = completionBlock;
-  _foregroundRipple.point = point;
-  [self reset:YES];
+  [self resetBottomInk:YES toPoint:point completion:completionBlock];
 }
 
 #pragma mark - MDCInkLayerRippleDelegate
 
-- (void)animationDidStop:(CAAnimation *)anim finished:(BOOL)finished {
-  if (finished) {
-    if (_spreadCompletionBlock) {
-      _spreadCompletionBlock();
-    }
-    if (_evaporateCompletionBlock) {
-      _evaporateCompletionBlock();
-    }
-    if (_evaporateToPointCompletionBlock) {
-      _evaporateToPointCompletionBlock();
-    }
+- (void)animationDidStop:(CAAnimation *)anim
+              shapeLayer:(CAShapeLayer *)shapeLayer
+                finished:(BOOL)finished {
+
+  if ([shapeLayer isKindOfClass:[MDCInkLayerRipple class]]) {
+    [(MDCInkLayerRipple *)shapeLayer clearAnimations];
+  }
+  if ([shapeLayer isMemberOfClass:[MDCInkLayerForegroundRipple class]]) {
+    [self.foregroundRipples removeObject:(MDCInkLayerForegroundRipple *)shapeLayer];
+    self.unexitedForegroundRippleIndex--;
```

Comment:
It feels like these `unexited` vars could be replaced by using the size of the array instead. Is there a strong need for these vars that perhaps I'm not seeing?